c_src/quicer_nif.c

```diff
@@ -1072,124 +1066,72 @@ openLib(ErlNifEnv *env, __unused_parm__ int argc, const ERL_NIF_TERM argv[])
   ERL_NIF_TERM lttngLib = argv[0];
   char lttngPath[PATH_MAX] = { 0 };
 
-  if (isLibOpened)
+  pthread_mutex_lock(&MsQuicLock);
+  if (MsQuic)
     {
+      // already opened
       TP_NIF_3(skip, 0, 2);
-      return SUCCESS(res);
+      res = SUCCESS(res);
+      goto exit;
     }
 
-  // @todo external call for static link
-  CxPlatSystemLoad();
-  MsQuicLibraryLoad();
-
   //
   // Open a handle to the library and get the API function table.
   //
   if (QUIC_FAILED(status = MsQuicOpen2(&MsQuic)))
     {
-      isLibOpened = false;
-      return ERROR_TUPLE_3(ATOM_OPEN_FAILED, ATOM_STATUS(status));
+      MsQuic = NULL;
+      res = ERROR_TUPLE_3(ATOM_OPEN_FAILED, ATOM_STATUS(status));
+      goto exit;
     }
 
-  isLibOpened = true;
   TP_NIF_3(success, 0, 2);
 
-  res = ATOM_TRUE;
+  res = SUCCESS(ATOM_TRUE);
 
   if (enif_get_string(env, lttngLib, lttngPath, PATH_MAX, ERL_NIF_LATIN1))
     {
       // loading lttng lib is optional, ok to fail
       if (dlopen(lttngPath, (unsigned)RTLD_NOW | (unsigned)RTLD_GLOBAL))
         {
-          res = ATOM_DEBUG;
+          res = SUCCESS(ATOM_DEBUG);
         }
     }
 
-  return SUCCESS(res);
+exit:
+  pthread_mutex_unlock(&MsQuicLock);
+  return res;
 }
 
 static ERL_NIF_TERM
 closeLib(__unused_parm__ ErlNifEnv *env,
          __unused_parm__ int argc,
          __unused_parm__ const ERL_NIF_TERM argv[])
 {
-  if (isLibOpened && MsQuic)
+  pthread_mutex_lock(&MsQuicLock);
+  if (MsQuic)
     {
-      // @todo ensure registration is closed first
-      //
-      TP_NIF_3(do_close, 0, isLibOpened);
-      MsQuicClose(MsQuic);
-      isLibOpened = false;
-    }
-
-  return ATOM_OK;
-}
-
-/*
-** For global registration only
-*/
-static ERL_NIF_TERM
-registration(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
-{
-  QUIC_STATUS status = QUIC_STATUS_SUCCESS;
-  ERL_NIF_TERM profile = argv[0];
+      TP_NIF_3(do_close, MsQuic, 0);
 
-  if (isRegistered || !isLibOpened)
-    {
-      return ERROR_TUPLE_2(ATOM_BADARG);
-    }
-
-  if (argc == 1)
-    {
-      if (IS_SAME_TERM(profile, ATOM_QUIC_EXECUTION_PROFILE_LOW_LATENCY))
-        {
-          GRegConfig.ExecutionProfile = QUIC_EXECUTION_PROFILE_LOW_LATENCY;
-        }
-      else if (IS_SAME_TERM(profile,
-                            ATOM_QUIC_EXECUTION_PROFILE_TYPE_MAX_THROUGHPUT))
+      pthread_mutex_lock(&GRegLock);
```

Comment:
Newbie question: are there other places that may try to take this lock and `MsQuicLock` after that?

Reply:
I should add doc that this GRegLock is only for library upgrade/load unload or testing. Shouldn't be used by application.